finality-aleph/src/finalization.rs

```diff
@@ -77,194 +80,133 @@ where
-                    100u64,
-                );
-                assert!(maybe_data.is_none());
+            // At this point we know that data validation went through, which means that the blocks are available
+            // and the whole `new_data` is honest and correct. Note that there might be cases in which some
+            // other node actually failed data validation for `new_data`. This could have happened in two cases:
+            // 1) Data availability let new_data through even though the `proposal` `new_data.branch.last()` was not available
+            // because it could infer that the `proposal` couldn't be finalized anyway (some conflicting block was finalized).
+            // 2) The whole `new_data.branch` was available and data availability let it through, but somehow between that moment and
+            // the moment we obtained `new_data` here, some blocks from the branch were pruned (this, most likely, can happen only with pruning
+            // is enabled).
+            //
+            // Both these cases are handled properly because if data validation does not fail, then the `last_finalized` is guaranteed
+            // to be a block conflicting with `new_data.branch.last()` and hence we will output `None` anyway.
+
+            let proposed_block = proposal.top_block();
+
+            // Early skip for optimization
+            if proposed_block.hash == last_finalized.hash
+                || proposed_block.num <= last_finalized.num
+            {
+                return None;
             }
-        }
 
-        let extra_children: Vec<_> = blocks
-            .iter()
-            .map(|hash| {
-                let mut builder = client
-                    .new_block_at(&BlockId::Hash(*hash), Default::default(), false)
-                    .unwrap();
-                // Add a dummy extrinsic to make the block distinct from the one on chain
-                builder
-                    .push(Extrinsic::AuthoritiesChange(Vec::new()))
-                    .unwrap();
-                let block = builder.build().unwrap().block;
-                let hash = block.header.hash();
-                futures::executor::block_on(client.import(BlockOrigin::Own, block)).unwrap();
-                hash
-            })
-            .collect();
+            // Iterate ancestors of `proposed_hash` until reaching a block with number <= last_finalized_number
+            // in order to check if proposed_block.hash is an ancestor of last_finalized
+            let mut hash = proposed_block.hash;
+            loop {
+                let header = match client
+                    .header(BlockId::Hash(hash))
+                    .expect("client must respond query")
+                {
+                    Some(header) => header,
+                    None => {
+                        error!(target: "afa", "No header for hash {}", hash);
+                        // We must panic here, as this is a major safety failure.
```

Comment:
don't we have the warranty that it is available? by the check `is_correct_and_available`?

Reply:
Then we kinda should have panicked there... maybe? I think this ties into my proposal with more specific correctness results in another comment – with these we could match on them, and `Finalizable` could even contain the header (perhaps making it more generic over the availability provider, with an associated type there... I'm probably overengineering a bit now though <_<").

Reply:
Hopeless forks are filtered out at `is_proposal_available_and_correct`. So this here means that this block was there during `is_proposal_available_and_correct` but now, for some reason it is not here anymore. I don't know why would that happen (substrate could randomly prune this block or something), but we can either ignore this and hope that this proposal is dummy (risking that other nodes could finalize a different block), or explicitly panic. I chose the latter.

Reply:
My point was you could have a version of `is_proposal_available_and_correct` that would return the block/header/whatever is needed, eliminating the need for a panic (or perhaps making the situation that requires a panic clearer). If the block can disappear between then and now, it can also disappear between this check and the actual finalization happening, so that solution is at least not worse.

Reply:
All this mess is now gone after applying the "other" comment.